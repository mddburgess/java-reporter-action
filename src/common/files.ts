import * as glob from '@actions/glob';

export async function findFiles(searchPaths: string[]) {
    const globber = await glob.create(searchPaths.join('\n'));
    return await globber.glob();
}

export async function findFile(searchPath: string) {
    const paths = await findFiles([searchPath]);
    return paths[0];
}

export async function findRelativePath(searchPath: string) {
    const globber = await glob.create(searchPath);
    const searchRoot = globber.getSearchPaths()[0];
    const absolutePaths = await globber.glob();
    return absolutePaths[0] && absolutePaths[0].slice(searchRoot.length + 1);
}
