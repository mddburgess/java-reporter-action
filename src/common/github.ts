export interface Annotation {
    path: string,
    start_line: number,
    end_line: number,
    start_column?: number,
    end_column?: number,
    annotation_level: AnnotationLevel,
    title?: string,
    message: string,
    raw_details?: string
}

export type AnnotationLevel =
    | 'notice'
    | 'warning'
    | 'failure';
