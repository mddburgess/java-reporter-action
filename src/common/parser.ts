import * as core from "@actions/core";
import * as fs from "fs";
import { Saxophone } from "saxophone-ts";
import {
  CDATANode,
  TagCloseNode,
  TagOpenNode,
  TextNode,
} from "saxophone-ts/dist/types/src/static/nodes";

export default abstract class ReportParser<T> {
  private readonly context: string[] = ["root"];
  private readonly parser = new Saxophone()
    .on("tagOpen", (tag) => this.handleTagOpen(tag))
    .on("tagClose", (tag) => this.handleTagClose(tag))
    .on("text", (tag) => this.handleText(tag))
    .on("cdata", (tag) => this.handleText(tag));

  protected constructor(protected readonly report: T, private readonly reportPath: string) {}

  read() {
    try {
      const xml = fs.readFileSync(this.reportPath, { encoding: "utf-8" });
      this.parser.parse(xml);
      return this.report;
    } catch (error) {
      core.warning(`Failed to read report: ${this.reportPath}`);
      return undefined;
    }
  }

  protected getContext() {
    return this.context[0];
  }

  private handleTagOpen(tag: TagOpenNode) {
    this.onTagOpen(tag);
    if (tag.isSelfClosing) {
      this.onTagClose(tag);
    } else {
      this.context.unshift(tag.name);
    }
  }

  private handleTagClose(tag: TagCloseNode) {
    this.context.shift();
    this.onTagClose(tag);
  }

  private handleText(tag: TextNode | CDATANode) {
    this.onText(tag);
  }

  protected abstract onTagOpen(tag: TagOpenNode): void;
  protected abstract onTagClose(tag: TagCloseNode): void;
  protected abstract onText(tag: TextNode | CDATANode): void;
}
