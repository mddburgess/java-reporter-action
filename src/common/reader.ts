import * as core from '@actions/core';
import * as fs from 'fs';
import {Saxophone} from 'saxophone-ts';
import {CDATANode, TagCloseNode, TagOpenNode, TextNode} from 'saxophone-ts/dist/types/src/static/nodes';

abstract class ReportReader<T> {

    private readonly parser = new Saxophone()
        .on('tagOpen', tag => this.handleTagOpen(tag))
        .on('tagClose', tag => this.handleTagClose(tag))
        .on('text', tag => this.handleText(tag))
        .on('cdata', tag => this.handleText(tag));
    private readonly context: string[] = ['root'];
    protected report?: T;

    public readReport(reportPath: string): T | undefined {
        try {
            const xml = fs.readFileSync(reportPath, {encoding: 'utf-8'});
            this.parser.parse(xml);
            return this.report;
        } catch (error) {
            core.warning(`Failed to read report: ${reportPath}`);
            console.log(error);
            return undefined;
        }
    }

    protected getContext(): string {
        return this.context[0];
    }

    private handleTagOpen(tag: TagOpenNode): void {
        this.onTagOpen(tag);
        if (tag.isSelfClosing) {
            this.onTagClose(tag);
        } else {
            this.context.unshift(tag.name);
        }
    }

    private handleTagClose(tag: TagCloseNode): void {
        this.context.shift();
        this.onTagClose(tag);
    }

    private handleText(tag: TextNode | CDATANode): void {
        this.onText(tag);
    }

    protected abstract onTagOpen(tag: TagOpenNode): void;
    protected abstract onTagClose(tag: TagCloseNode): void;
    protected abstract onText(tag: TextNode | CDATANode): void;
}

export default ReportReader;
