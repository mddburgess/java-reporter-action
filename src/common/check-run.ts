import * as core from '@actions/core';
import * as github from '@actions/github';
import {Annotation} from './github';

class CheckRun {

    private readonly token;
    private readonly octokit;

    private readonly name;
    private checkRunId?: number;

    constructor(name: string) {
        this.token = core.getInput('github-token', {required: true});
        this.octokit = github.getOctokit(this.token);
        this.name = name;
    }

    public async conclude(request: CheckRunRequest) {
        if (request.output) {
            await this.createOrUpdateCheckRun({
                status: 'completed',
                conclusion: request.conclusion,
                output: {
                    title: request.output.title,
                    summary: request.output.summary
                }
            });
            await this.annotate(request.output);
        } else {
            await this.createOrUpdateCheckRun({
                status: 'completed',
                conclusion: request.conclusion
            });
        }
    }

    private async annotate(output: CheckOutput) {
        if (!output.annotations || output.annotations.length === 0) {
            return;
        }
        for (let i = 0; i < output.annotations.length; i += 50) {
            await this.updateCheckRun({
                output: {
                    title: output.title,
                    summary: output.summary,
                    annotations: output.annotations.slice(i, i + 50)
                }
            });
        }
    }

    private async createOrUpdateCheckRun(request: CheckRunRequest) {
        if (this.checkRunId === undefined) {
            await this.createCheckRun(request);
        } else {
            await this.updateCheckRun(request);
        }
    }

    private async createCheckRun(request: CheckRunRequest) {
        core.debug(`Creating ${this.name} check run...`);
        const response = await this.octokit.checks.create({
            ...github.context.repo,
            name: this.name,
            head_sha: this.resolveHeadSha(),
            ...request
        });
        this.checkRunId = response.data.id;
    }

    private async updateCheckRun(request: CheckRunRequest) {
        if (this.checkRunId === undefined) {
            throw Error('Cannot update a check run before creating it');
        }

        core.debug(`Updating ${this.name} check run...`);
        await this.octokit.checks.update({
            ...github.context.repo,
            check_run_id: this.checkRunId,
            ...request
        });
    }

    private resolveHeadSha() {
        return github.context.payload.pull_request
            ? github.context.payload.pull_request.head.sha
            : github.context.sha;
    }
}

interface CheckRunRequest {
    status?: CheckStatus,
    conclusion?: CheckConclusion,
    output?: CheckOutput
}

type CheckStatus =
    | 'queued'
    | 'in_progress'
    | 'completed';

type CheckConclusion =
    | 'success'
    | 'failure'
    | 'neutral'
    | 'cancelled'
    | 'skipped'
    | 'timed_out'
    | 'action_required';

interface CheckOutput {
    title: string,
    summary: string,
    text?: string,
    annotations?: Annotation[];
}

export default CheckRun;
