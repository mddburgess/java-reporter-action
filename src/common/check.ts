import * as core from '@actions/core';
import * as github from '@actions/github';
import * as glob from '@actions/glob';
import path from 'path';
import {getInput} from '@actions/core';

abstract class Check {

    private static readonly workspacePath = process.env.GITHUB_WORKSPACE || '';

    private readonly reportType: string;
    private readonly checkCondition: CheckCondition;
    private checkRunId?: number;

    protected constructor(reportType: string) {
        this.reportType = reportType;
        this.checkCondition = this.resolveCheckCondition();
    }

    private resolveCheckCondition() {
        const condition = getInput(this.reportType);
        switch (condition) {
            case 'required':
            case 'expected':
            case 'disabled':
                return CheckCondition[condition];
            case undefined:
                return CheckCondition.autodetect;
            default:
                core.warning(`Input '${this.reportType}' is invalid - must be one of ['required','expected','disabled']. Defaulting to 'expected'.`);
                return CheckCondition.expected;
        }
    }

    public async run() {
        const reportPaths = await this.findReports();

        const token = core.getInput('github-token', {required: true});
        const octokit = github.getOctokit(token);

        if (reportPaths.length > 0 || this.checkCondition >= CheckCondition.expected) {
            const response = await octokit.checks.create({
                ...github.context.repo,
                name: this.reportType,
                head_sha: github.context.payload.pull_request
                    ? github.context.payload.pull_request.head.sha
                    : github.context.sha,
                status: 'completed',
                conclusion: reportPaths.length ? 'success' :
                    this.checkCondition === CheckCondition.required ? 'failure' : 'skipped'
            });
            this.checkRunId = response.data.id;
        }

        core.info(`${this.reportType} check finished.`);
    }

    private async findReports() {
        core.startGroup(`Searching for ${this.reportType} reports...`);

        const searchPaths = this.reportSearchPaths();
        searchPaths.forEach(searchPath => core.info(searchPath));
        core.endGroup();

        const globber = await glob.create(searchPaths.join('\n'));
        const reportPaths = await globber.glob();

        core.startGroup(`Found ${reportPaths.length} ${this.reportType} reports.`);
        reportPaths.forEach(reportPath => core.info(path.relative(Check.workspacePath, reportPath)));
        core.endGroup();

        return reportPaths;
    }

    protected abstract reportSearchPaths(): string[];
}

enum CheckCondition {
    disabled,
    autodetect,
    expected,
    required
}

export default Check;
