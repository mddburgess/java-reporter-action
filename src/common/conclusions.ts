import { AnnotationSummary } from "../github/check-run";
import { CheckConclusion } from "../github/types";

const failureOnWarnings = (result: AnnotationSummary): CheckConclusion => {
  if (result.failures + result.warnings > 0) {
    return "failure";
  }
  return "success";
};

const neutralOnWarnings = (result: AnnotationSummary): CheckConclusion => {
  if (result.failures > 0) {
    return "failure";
  }
  if (result.warnings > 0) {
    return "neutral";
  }
  return "success";
}

export {
  failureOnWarnings,
  neutralOnWarnings
};
