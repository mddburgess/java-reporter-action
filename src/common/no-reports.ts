import CheckResult from "../check/result";
import { CheckAnnotation, CheckConclusion } from "../github/types";
import { RunCondition } from "../check";

export default class NoReportsResult extends CheckResult {
  constructor(
    private readonly friendlyName: string,
    private readonly runCondition: RunCondition,
    private readonly searchPaths: string[]
  ) {
    super();
  }

  shouldCompleteCheck(): boolean {
    return this.runCondition >= RunCondition.expected;
  }

  get conclusion(): CheckConclusion {
    return this.runCondition === RunCondition.required ? "failure" : "skipped";
  }

  get title(): string {
    return "No reports found";
  }

  get summary(): string {
    const runConditionName = RunCondition.required ? "required" : "expected";
    return `The ${this.friendlyName} check is ${runConditionName}, but no ${this.friendlyName} reports were found.`;
  }

  get text(): string | undefined {
    return ["### Search paths", "```sh", ...this.searchPaths, "```"].join("\n");
  }

  get annotations(): CheckAnnotation[] | undefined {
    return undefined;
  }
}
