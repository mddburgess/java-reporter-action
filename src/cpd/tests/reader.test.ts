import {findFile} from '../../common/files';
import CpdReportReader from '../reader';

describe('CPD report reader', () => {

    it('can read a PMD report', async () => {
        const reportPath = await findFile('**/cpd.xml');
        const report = new CpdReportReader().readReport(reportPath);
        expect(report).toMatchSnapshot();
    });
});
