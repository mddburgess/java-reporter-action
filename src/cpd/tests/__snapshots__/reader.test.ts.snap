// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CPD report reader can read a PMD report 1`] = `
Object {
  "duplications": Array [
    Object {
      "files": Array [
        Object {
          "endColumn": 10,
          "endLine": 106,
          "path": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/cpd/ErrorProne.java",
          "startColumn": 33,
          "startLine": 46,
        },
        Object {
          "endColumn": 10,
          "endLine": 111,
          "path": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/ErrorProne.java",
          "startColumn": 58,
          "startLine": 51,
        },
      ],
      "lines": 61,
      "tokens": 163,
    },
    Object {
      "files": Array [
        Object {
          "endColumn": 5,
          "endLine": 143,
          "path": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/cpd/ErrorProne.java",
          "startColumn": 5,
          "startLine": 104,
        },
        Object {
          "endColumn": 5,
          "endLine": 52,
          "path": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/ErrorProne.java",
          "startColumn": 13,
          "startLine": 13,
        },
      ],
      "lines": 40,
      "tokens": 154,
    },
    Object {
      "files": Array [
        Object {
          "endColumn": 5,
          "endLine": 47,
          "path": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/cpd/ErrorProne.java",
          "startColumn": 13,
          "startLine": 11,
        },
        Object {
          "endColumn": 5,
          "endLine": 145,
          "path": "/Users/mike/Projects/java-reporter-action/example/src/main/java/org/example/pmd/ErrorProne.java",
          "startColumn": 5,
          "startLine": 109,
        },
      ],
      "lines": 37,
      "tokens": 135,
    },
  ],
}
`;
